### PR TITLE
PSC-4 RequestManager

### PR DESCRIPTION
**Pull request by BohdanDatskiv**
https://trello.com/c/SgBGZr1k

### CODE REVIEW

File: Uni/Business Logic/RequestManager/NetworkResponse.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+enum NetworkResponse: Int, Error {
+    case badRequest
+    case failed
+    case noData
+    case unableToDecode
+    case unauthorized = 401
+    case forbidden = 403
+    case notFound = 404
+    case internalServerError = 500
+    case incorrectHTTPMethod
```
Comment (Alex-Chaku) [inline Uni/Business Logic/RequestManager/NetworkResponse.swift]: Remove raw value for enum, and add parameters where it's needed like incorrectHTTPMethod(value: ) to make errors more describing. Also, add computed property "var userFriendlyText: ErrorInfo { get }" where "ErrorInfo" it's or tuple or struct that contains "message" and "description". (can be the same text for multiple errors)

File: Uni/Business Logic/RequestManager/RequestManager.swift [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    func run<T: Codable>(request: UrlRequestConvertible) -> Promise<T> {
+        return Promise { seal in
+            guard let request = try? request.asURLRequest() else {
+                seal.reject(NetworkResponse.badRequest)
+                return
+            }
+            currentRequests.insert(request)
+            queue.async {
+                self.checkHTTPMethod(in: request).then {
+                    self.retry {
```
Comment (Alex-Chaku) [inline Uni/Business Logic/RequestManager/RequestManager.swift]: Retry is implemented good, despite there should be strategy for that. You should ask if it's needed to retry

File: Uni/Business Logic/RequestManager/RequestManager.swift [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    private let queue: DispatchQueue
+    var currentRequests = Set<URLRequest>()
+
+    /// Init
+    init(queue: DispatchQueue? = nil) {
+        if let queue = queue {
+            self.queue = queue
+        } else {
+            self.queue = DispatchQueue(label: "OwnQueue", qos: .background)
```
Comment (Alex-Chaku) [inline Uni/Business Logic/RequestManager/RequestManager.swift]: what's name "OwnQueue", rename to something more describing, like "uni_request_manager_queue"

File: Uni/Business Logic/RequestManager/Models/ResponseContainer.swift [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            if let value = try? dynamicContainer.decodeIfPresent(T.self, forKey: key) {
+                genericResult = value
+            }
+        }
+        if let genericResult = genericResult {
+            result = genericResult
+        } else {
+            throw NetworkResponse.noData
+        }
+    }
```
Comment (Alex-Chaku) [inline Uni/Business Logic/RequestManager/Models/ResponseContainer.swift]: Nice👍🏻

File: Uni/Business Logic/RequestManager/Protocols/URLRequestConvertible.swift [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+//  Uni
+//
+//  Created by Bohdan Datskiv on 8/27/19.
+//
+
+import Foundation
+
+protocol UrlRequestConvertible {
+    func asURLRequest() throws -> URLRequest
+}
```
Comment (Alex-Chaku) [inline Uni/Business Logic/RequestManager/Protocols/URLRequestConvertible.swift]: We finally did it!!! No we have a separate file with "UrlRequestConvertible" protocol, it took only 2 weeks😂

File: Uni/Business Logic/RequestManager/RequestManager.swift [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+    private func decode<T: Codable>(data: Data) -> Promise<T> {
+        return .just(try JSONDecoder().decode(T.self, from: data))
+    }
+
+    private func validateResponse(_ response: URLResponse) -> Promise<Void> {
+        return Promise { seal in
+            if let response = response as? HTTPURLResponse {
+                if 200..<300 ~= response.statusCode {
+                    seal.fulfill(())
```
Comment (Alex-Chaku) [inline Uni/Business Logic/RequestManager/RequestManager.swift]: Fix crash, you can't call seal multiple times

File: Uni/Business Logic/RequestManager/NetworkResponse.swift [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+//  NetworkResponse.swift
+//  Uni
+//
+//  Created by Bohdan Datskiv on 8/28/19.
+//
+
+import Foundation
+
+enum NetworkResponse: Int, Error {
+    case badRequest
```
[the remark this answers is not in the document]
Comment (BohdanDatskiv) [reply Uni/Business Logic/RequestManager/NetworkResponse.swift]: fixed

File: Uni/Business Logic/RequestManager/RequestManager.swift (hunk 2 above)
Comment (BohdanDatskiv) [reply Uni/Business Logic/RequestManager/RequestManager.swift]: fixed

File: Uni/Business Logic/RequestManager/NetworkResponse.swift (hunk 1 above)
Comment (BohdanDatskiv) [reply Uni/Business Logic/RequestManager/NetworkResponse.swift]: fixed

File: Uni/Business Logic/RequestManager/RequestManager.swift [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            } else if let error = NetworkResponse(rawValue: response.statusCode) {
+                seal.reject(error)
+            }
+        }
+    }
+
+    private func checkHTTPMethod(in request: URLRequest) -> Promise<Void> {
+        return Promise { seal in
+            guard let method = request.httpMethod, method == "GET" else {
+                seal.reject(NetworkResponse.incorrectHTTPMethod)
```
[the remark this answers is not in the document]
Comment (BohdanDatskiv) [reply Uni/Business Logic/RequestManager/RequestManager.swift]: fixed

[the remark this answers is not in the document]
Comment (BohdanDatskiv) [reply Uni/Business Logic/RequestManager/RequestManager.swift]: fixed

File: Uni/Business Logic/RequestManager/RequestManager.swift (hunk 3 above)
Comment (BohdanDatskiv) [reply Uni/Business Logic/RequestManager/RequestManager.swift]: done

File: Uni/Business Logic/RequestManager/Models/ResponseContainer.swift (hunk 4 above)
Comment (BohdanDatskiv) [reply Uni/Business Logic/RequestManager/Models/ResponseContainer.swift]: done

File: Uni/Business Logic/RequestManager/RequestManager.swift (hunk 6 above)
Comment (BohdanDatskiv) [reply Uni/Business Logic/RequestManager/RequestManager.swift]: fixed

**Pull request merged**